lib/services/unified_resource.go

```diff
@@ -204,116 +207,275 @@ func (c *UnifiedResourceCache) getSortTree(sortField string) (*btree.BTreeG[*ite
+				if !ok {
+					return true
+				}
+
+				if len(kinds) == 0 {
+					return yield(iteratedItem{key: item.Key, resource: r}, nil)
+				}
+
+				if c.itemKindMatches(r, kindsMap) {
+					return yield(iteratedItem{key: item.Key, resource: r}, nil)
+				}
+
 				return true
+			})
+
+			return nil
+		})
+		if err != nil {
+			yield(iteratedItem{}, err)
+		}
+	}
+}
+
+// IterateResources iterates over all resources from the start key that match
+// one of the provided kinds. If no kinds are provided, resources of all supported
+// kinds are returned.
+func (c *UnifiedResourceCache) IterateResources(ctx context.Context, start string, sortBy types.SortBy, kinds ...string) iter.Seq2[types.ResourceWithLabels, error] {
+	return func(yield func(types.ResourceWithLabels, error) bool) {
+		for item, err := range c.iterateItems(ctx, start, sortBy, kinds...) {
+			if err != nil {
+				yield(nil, err)
+				return
 			}
 
-			// do we have all we need? set nextKey and stop iterating
-			// we do this after the matchFn to make sure they have access to the "next" node
-			if req.Limit > 0 && len(res) >= int(req.Limit) {
-				nextKey = item.Key.String()
-				return false
+			if !yield(item.resource.CloneResource(), nil) {
+				return
 			}
-			res = append(res, resourceFromMap)
-			return true
-		})
-		return iteratorErr
-	})
-	if err != nil {
-		return nil, "", trace.Wrap(err)
+		}
 	}
+}
+
+// IterateNodes iterates over all cached nodes starting from the provided key.
+func (c *UnifiedResourceCache) IterateNodes(ctx context.Context, start string, sortBy types.SortBy) iter.Seq2[types.Server, error] {
+	return iterateUnifiedResourceCache(ctx, c, start, sortBy, types.KindNode, types.Server.DeepCopy)
+}
+
+// IterateAppServers iterates over all cached app servers starting from the provided key.
+func (c *UnifiedResourceCache) IterateAppServers(ctx context.Context, start string, sortBy types.SortBy) iter.Seq2[types.AppServer, error] {
+	return iterateUnifiedResourceCache(ctx, c, start, sortBy, types.KindApp, types.AppServer.Copy)
+}
 
-	if len(res) == backend.DefaultRangeLimit {
-		c.logger.WarnContext(ctx, "Range query hit backend limit. (this is a bug!)",
-			"start_key", startKey,
-			"range_limit", backend.DefaultRangeLimit,
-		)
+// IterateDatabaseServers iterates over all cached database servers starting from the provided key.
+func (c *UnifiedResourceCache) IterateDatabaseServers(ctx context.Context, start string, sortBy types.SortBy) iter.Seq2[types.DatabaseServer, error] {
+	return iterateUnifiedResourceCache(ctx, c, start, sortBy, types.KindDatabase, types.DatabaseServer.Copy)
+}
+
+// IterateKubernetesServers iterates over all cached Kubernetes servers starting from the provided key.
+func (c *UnifiedResourceCache) IterateKubernetesServers(ctx context.Context, start string, sortBy types.SortBy) iter.Seq2[types.KubeServer, error] {
+	return iterateUnifiedResourceCache(ctx, c, start, sortBy, types.KindKubernetesCluster, types.KubeServer.Copy)
```

Comment:
if im reading this correctly, this only populates the `kindMap` in `iterateItems` with `KubernetesCluster` so if the request type is KindKubeServer, it wont be present in the `kindMap` so `c.itemKindMatches` wont contain the proper type per the request. we'll have to do some shenanigans to add the correct "container" and contained types for the servers